File: sleepsense-backend/src/main/java/com/sleepsense/pipeline/SensorPipeline.java
package com.sleepsense.pipeline;

import com.sleepsense.alert.AlertDispatcher;
import com.sleepsense.alert.AlertNotification;
import com.sleepsense.analysis.ThresholdChecker;
import com.sleepsense.analysis.ThresholdResult;
import com.sleepsense.config.ThresholdConfig;
import com.sleepsense.model.SensorData;
import com.sleepsense.service.SensorService;
import com.sleepsense.service.ThresholdSettingsService;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.List;

/**
 * getSensorData() -&gt; checkThreshold() -&gt; sendAlert()
 *
 * <p>Test Plan reference: ITC-01 (Test-SensorPipeline). Normal data flows
 * through without triggering an alert; critical data triggers exactly one
 * {@code sendAlert} call, which in turn emits one notification per critical
 * factor.
 */
@Component
@RequiredArgsConstructor
public class SensorPipeline {

    private final SensorService sensorService;
    private final ThresholdChecker thresholdChecker;
    private final AlertDispatcher alertDispatcher;
    private final ThresholdSettingsService thresholdSettingsService;

    /** What one pass of the pipeline produced. */
    @Getter
    public static class Outcome {
        private final SensorData reading;
        private final ThresholdResult result;
        private final List<AlertNotification> notifications;
        private final boolean alertSent;

        public Outcome(SensorData reading, ThresholdResult result,
                       List<AlertNotification> notifications, boolean alertSent) {
            this.reading = reading;
            this.result = result;
            this.notifications = notifications;
            this.alertSent = alertSent;
        }
    }

    public Outcome run(String deviceId) {
        SensorData reading = sensorService.getSensorData(deviceId);
        ThresholdConfig effective = thresholdSettingsService.getEffective(deviceId);
        ThresholdResult result = thresholdChecker.checkThreshold(reading, effective);

        if (!result.isCritical()) {
            return new Outcome(reading, result, Collections.emptyList(), false);
        }

        List<AlertNotification> notifications =
                alertDispatcher.sendAlert(deviceId, reading, result);
        return new Outcome(reading, result, notifications, true);
    }
}

File: lib/core/pipeline/sensor_pipeline.dart
import '../models/sensor_reading.dart';
import '../models/threshold_config.dart';
import '../services/alert_service.dart';
import '../services/sensor_service.dart';
import '../utils/threshold_checker.dart';

/// What one pass of the pipeline produced.
class PipelineOutcome {
  final SensorReading reading;
  final ThresholdResult result;

  /// The notifications that went out. Empty when nothing was critical, in which
  /// case [alertSent] is false and `sendAlert` was never called.
  final List<AlertNotification> notifications;
  final bool alertSent;

  const PipelineOutcome({
    required this.reading,
    required this.result,
    required this.notifications,
    required this.alertSent,
  });
}

/// getSensorData() -> checkThreshold() -> sendAlert()
///
/// Test Plan reference: ITC-01 (Test-SensorPipeline). Normal data flows through
/// without triggering an alert; critical data triggers exactly one `sendAlert`
/// call, which in turn emits one notification per critical factor.
class SensorPipeline {
  final SensorService sensorService;
  final AlertDispatcher alertDispatcher;
  final ThresholdConfig config;

  const SensorPipeline({
    required this.sensorService,
    required this.alertDispatcher,
    this.config = ThresholdConfig.defaults,
  });

  Future<PipelineOutcome> run() async {
    final reading = await sensorService.getSensorData();
    final result = checkThreshold(reading, config: config);

    if (!result.isCritical) {
      return PipelineOutcome(
        reading: reading,
        result: result,
        notifications: const [],
        alertSent: false,
      );
    }

    final notifications = await alertDispatcher.sendAlert(reading, result);
    return PipelineOutcome(
      reading: reading,
      result: result,
      notifications: notifications,
      alertSent: true,
    );
  }
}
